### ISSUE TITLE
Some modules may not involve the back propagation?

### ISSUE THREAD

**Issue by blackantywj** [sundongwei/MaskApproxNet#2]
When I run the command:

`
python scripts/segmentation_train.py --image_size 256 --num_channels 128 --class_cond False --num_res_blocks 2 --num_heads 1 --learn_sigma True --use_scale_shift_norm False --attention_resolutions 16 --diffusion_steps 1000 --noise_schedule linear --rescale_learned_sigmas False --rescale_timesteps False --lr 1e-4 --batch_size 8
`

I encounter the following error: 
`
Exception has occurred: RuntimeError
Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 0: module.position_encoding.embedding_1D.weight, module.transformer.layers.5.fc_alpha3.bias, module.transformer.layers.5.fc_alpha3.weight, module.transformer.layers.5.fc_alpha2.bias, module.transformer.layers.5.fc_alpha2.weight, module.transformer.layers.5.fc_alpha1.bias, module.transformer.layers.5.fc_alpha1.weight, module.transformer.layers.4.fc_alpha3.bias, module.transformer.layers.4.fc_alpha3.weight, module.transformer.layers.4.fc_alpha2.bias, module.transformer.layers.4.fc_alpha2.weight, module.transformer.layers.4.fc_alpha1.bias, module.transformer.layers.4.fc_alpha1.weight, module.transformer.layers.1.fc_alpha1.bias, module.transformer.layers.1.fc_alpha1.weight, module.transformer.layers.0.fc_alpha3.bias, module.transformer.layers.0.fc_alpha3.weight, module.transformer.layers.0.fc_alpha2.bias, module.transformer.layers.0.fc_alpha2.weight, module.transformer.layers.0.fc_alpha1.bias, module.transformer.layers.0.fc_alpha1.weight, module.transformer.layers.1.fc_alpha2.weight, module.transformer.layers.1.fc_alpha2.bias, module.transformer.layers.1.fc_alpha3.weight, module.transformer.layers.1.fc_alpha3.bias, module.transformer.layers.2.fc_alpha1.weight, module.transformer.layers.2.fc_alpha1.bias, module.transformer.layers.2.fc_alpha2.weight, module.transformer.layers.2.fc_alpha2.bias, module.transformer.layers.2.fc_alpha3.weight, module.transformer.layers.2.fc_alpha3.bias, module.transformer.layers.3.fc_alpha1.weight, module.transformer.layers.3.fc_alpha1.bias, module.transformer.layers.3.fc_alpha2.weight, module.transformer.layers.3.fc_alpha2.bias, module.transformer.layers.3.fc_alpha3.weight, module.transformer.layers.3.fc_alpha3.bias
Parameter indices which did not receive grad for rank 0: 33 34 35 36 37 38 57 58 59 60 61 62 81 82 83 84 85 86 105 106 107 108 109 110 129 130 131 132 133 134 153 154 155 156 157 158 159
`

**Comment by blackantywj** [comment]
I have commented the following codes and ran the project successfully. 😊

line 106 of decoder.py

`self.fc_alpha1 = nn.Linear(d_model + d_model, d_model)
self.fc_alpha2 = nn.Linear(d_model + d_model, d_model)
self.fc_alpha3 = nn.Linear(d_model + d_model, d_model)
self.init_weights()`

line 59 of decoder.py
`self.embedding_1D = nn.Embedding(52, int(d_model))`


Can you please explain the function of these codes?

**Comment by blackantywj** [comment]
After repair all errors above, I trained the model following README without changing another parameters and found that the caption loss can not converge in LEVIR-MCI dataset. By the way, I notice that some hyperparameters in github are different from their values in the paper (such as lr and momentum). Can you give me some advices? Thank you!

**Comment by sundongwei** [comment]
> After repair all errors above, I trained the model following README without changing another parameters and found that the caption loss can not converge in LEVIR-MCI dataset. By the way, I notice that some hyperparameters in github are different from their values in the paper (such as lr and momentum). Can you give me some advices? Thank you!

Hi, thank you very much for your feedback and for testing our model on the LEVIR-MCI dataset. Regarding the caption loss not converging, this is a valuable observation. Models(especially, diffusion) of this complexity can be very sensitive to new data distributions, and it's possible this is related to the inherent nature of the model, which may require an extremely long training cycle to show stable convergence on a dataset like LEVIR-MCI, rather than being a specific error. You were also sharp to notice the hyperparameter differences (like lr); the GitHub values are often a robust starting point, while the paper's values might be from a specific, fine-tuned experiment. We suggest you first try extending the training time significantly to see if the loss eventually trends downward, or perhaps try applying the exact hyperparameter set mentioned in the paper, which might be better suited for this specific dataset. Thank you again for raising this helpful issue, and we welcome any updates on your progress.

**Comment by blackantywj** [comment]
I'm sorry for bothering you again! I train the model with more steps but the gradient is explosion as follows: 

---------------------------
| caption_loss | nan      |
| grad_norm    | 16.2     |
| loss         | 0.000243 |
| loss_cal     | 0.0072   |
| loss_cal_q0  | 0.000334 |
| loss_cal_q2  | 0.00441  |
| loss_cal_q3  | 0.0197   |
| loss_q0      | 0.000504 |
| loss_q2      | 0.000217 |
| loss_q3      | 3.23e-05 |
| mse_diff     | 0.00024  |
| mse_diff_q0  | 0.000498 |
| mse_diff_q2  | 0.000215 |
| mse_diff_q3  | 3.19e-05 |
| param_norm   | 584      |
| samples      | 1.29e+06 |
| step         | 1.61e+05 |
| total_loss   | nan      |
| vb           | 2.41e-06 |
| vb_q0        | 5.45e-06 |
| vb_q2        | 1.87e-06 |
| vb_q3        | 4.45e-07 |
---------------------------
---------------------------
| caption_loss | nan      |
| grad_norm    | 0.321    |
| loss         | 0.00055  |
| loss_cal     | 0.00348  |
| loss_cal_q0  | 0.000679 |
| loss_cal_q1  | 0.00109  |
| loss_cal_q2  | 0.0118   |
| loss_cal_q3  | 4.35e-08 |
| loss_q0      | 0.00113  |
| loss_q1      | 0.000163 |
| loss_q2      | 0.000314 |
| loss_q3      | 5.58e-05 |
| mse_diff     | 0.000537 |
| mse_diff_q0  | 0.0011   |
| mse_diff_q1  | 0.000161 |
| mse_diff_q2  | 0.000311 |
| mse_diff_q3  | 5.51e-05 |
| param_norm   | 584      |
| samples      | 1.29e+06 |
| step         | 1.61e+05 |
| total_loss   | nan      |
| vb           | 1.33e-05 |
| vb_q0        | 3.24e-05 |
| vb_q1        | 1.25e-06 |
| vb_q2        | 2.9e-06  |
| vb_q3        | 7.35e-07 |
---------------------------


Could you give me some suggestions?
